src/webgpu/shader/execution/evaluation_order.spec.ts

```diff
@@ -0,0 +1,480 @@
+    },
+    {
+      name: 'AllSE',
+      _body: 'return sub3(mul(&a, 10), mul(&a, 10), mul(&a, 10));',
+      _result: 20 - 200 - 2000,
+    },
+    {
+      name: 'MiddleNotSE',
+      _body: 'return sub3(mul(&a, 10), a, mul(&a, 10));',
+      _result: 20 - 20 - 200,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('index_accessor')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#array-access-expr')
+  .desc('Tests order of evaluation of index accessor expressions.')
+  .paramsSimple([
+    {
+      name: 'LeftSE', //
+      _body: 'return arr2D[mul(&a, 2)][a];',
+      _result: 4 * 10 + 4,
+    },
+    {
+      name: 'RightSE', //
+      _body: 'return arr2D[a][mul(&a, 2)];',
+      _result: 2 * 10 + 4,
+    },
+    {
+      name: 'BothSE',
+      _body: 'return arr2D[mul(&a, 2)][mul(&a, 2)];',
+      _result: 4 * 10 + 8,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('assignment')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#assignment')
+  .desc('Tests order of evaluation of assignment statements.')
+  .paramsSimple([
+    {
+      name: 'ToArray1D',
+      _body:
+        'arr1D_zero[mul(&a, 2)] = mul(&a, 2);' + //
+        'return arr1D_zero[8];',
+      _result: 4,
+    },
+    {
+      name: 'ToArray2D',
+      _body:
+        'arr2D_zero[mul(&a, 2)][mul(&a, 2)] = mul(&a, 2);' + //
+        'return arr2D_zero[8][16];',
+      _result: 4,
+    },
+    {
+      name: 'ToArray4D',
+      _body:
+        'arr3D_zero[mul(&a, 2)][mul(&a, 2)][mul(&a, 2)] = mul(&a, 2);' +
+        'return arr3D_zero[8][16][32];',
+      _result: 4,
+    },
+    {
+      name: 'ToArrayFromArray',
+      _body:
+        'arr2D_zero[4][8] = 123;' +
+        'arr1D_zero[mul(&a, 2)] = arr2D_zero[mul(&a, 2)][mul(&a, 2)];' +
+        'return arr1D_zero[16];',
+      _result: 123,
+    },
+    {
+      name: 'ToArrayIndexedByArrayIndexedBySE',
+      _body:
+        'var arr1 = arr1D_zero;' +
+        'var arr2 = arr1D_zero;' +
+        'arr2[8] = 3;' +
+        'arr1[arr2[mul(&a, 2)]] = mul(&a, 2);' +
+        'return arr1[3];',
+      _result: 4,
+    },
+    {
+      name: 'ToVec_BothSE',
+      _body:
+        'a = 0;' +
+        'vec4_zero[add(&a, 1)] = add(&a, 1);' + //
+        'return vec4_zero[2];',
+      _result: 1,
+    },
+    {
+      name: 'ToVec_LeftSE',
+      _body:
+        'vec4_zero[add(&a, 1)] = a;' + //
+        'return vec4_zero[3];',
+      _result: 2,
+    },
+    {
+      name: 'ToVec_RightSE',
+      _body:
+        'vec4_zero[a] = add(&a, 1);' + //
+        'return vec4_zero[3];',
+      _result: 3,
+    },
+  ])
+  .fn(t => run(t, t.params._body, t.params._result));
+
+g.test('type_constructor')
+  .specURL('https://gpuweb.github.io/gpuweb/wgsl/#type-constructor-expr')
+  .desc('Tests order of evaluation of type constructor expressions.')
+  .paramsSimple([
+    {
+      name: 'Struct',
+      _body:
+        'let r = S(mul(&a, 10), mul(&a, 10), mul(&a, 10));' + //
```

Comment:
Here's another example of where having different sub-expressions would make the tests stronger.

Reply:
Done.